File: src/third_iteration/anonymous_classes/Gesture.java
package third_iteration.anonymous_classes;

public interface Gesture {

	Gesture ROCK = new Gesture() {
		@Override
		public String againstRock() {
			return "Tie.";
		}

		@Override
		public String againstPaper() {
			return "Paper wins against rock.";
		}

		@Override
		public String againstScissors() {
			return "Scissors lose against rock.";
		}

		@Override
		public String compare(Gesture gesture) {
			return gesture.againstRock();
		}
	};
	Gesture PAPER = new Gesture() {
		@Override
		public String againstRock() {
			return "Rock loses against paper.";
		}

		@Override
		public String againstPaper() {
			return "Tie.";
		}

		@Override
		public String againstScissors() {
			return "Scissors win against paper.";
		}

		@Override
		public String compare(Gesture gesture) {
			return gesture.againstPaper();
		}
	};
	Gesture SCISSORS = new Gesture() {
		@Override
		public String againstRock() {
			return "Rock wins against scissors.";
		}

		@Override
		public String againstPaper() {
			return "Paper loses against scissors.";
		}

		@Override
		public String againstScissors() {
			return "Tie.";
		}

		@Override
		public String compare(Gesture gesture) {
			return gesture.againstScissors();
		}
	};

	String againstRock();
	String againstPaper();
	String againstScissors();
	String compare(Gesture gesture);
}

File: php7/src/third_iteration/anonymous_classes/Gesture.php
<?php

declare(strict_types = 1);

namespace third_iteration\anonymous_classes;

abstract class Gesture
{
    abstract function againstRock():string;

    abstract function againstPaper():string;

    abstract function againstScissors():string;

    abstract function compare(Gesture $gesture):string;

    public static function ROCK():Gesture
    {
        return new class() extends Gesture
        {
            public function againstRock():string
            {
                return "Tie.";
            }

            public function againstPaper():string
            {
                return "Paper wins against rock.";
            }

            public function againstScissors():string
            {
                return "Scissors lose against rock.";
            }

            public function compare(Gesture $gesture):string
            {
                return $gesture->againstRock();
            }
        };
    }

    public static function PAPER():Gesture
    {
        return new class() extends Gesture
        {
            public function againstRock():string
            {
                return "Rock loses against paper.";
            }

            public function againstPaper():string
            {
                return "Tie.";
            }

            public function againstScissors():string
            {
                return "Scissors win against paper.";
            }

            public function compare(Gesture $gesture):string
            {
                return $gesture->againstPaper();
            }
        };
    }

    public static function SCISSORS():Gesture
    {
        return new class() extends Gesture
        {
            public function againstRock():string
            {
                return "Rock wins against scissors.";
            }

            public function againstPaper():string
            {
                return "Paper loses against scissors.";
            }

            public function againstScissors():string
            {
                return "Tie.";
            }

            public function compare(Gesture $gesture):string
            {
                return $gesture->againstScissors();
            }
        };
    }
}